### ISSUE TITLE
Could Android device enable adb debug mode automatically  after factory reset?

### ISSUE THREAD

**Issue by Tierney0115** [openstf/stf#213]
We are considering allow user to do factory reset and after that keep the device online without human interface.
Could the devices just run a script after reboot to open the USB debug mode?  
Thanks.

**Comment by cemakkok** [comment]
Yeah, that is so important. Do you have an any idea @sorccu ?

**Comment by sorccu** [comment]
No, but then again I have not researched this topic very well.

**Comment by phhusson** [comment]
Here is the script I use to do that:
https://gist.github.com/phhusson/3d1e8833309dc34ee5502a409b80e292
You need to give it a blank userdata.img, or the userdata.img provided by your OEM,
This has been tested on various devices, ranging from 5.1 to 7.1, either AOSP-based, or MTK based.
For Spreadtrum-based, some firmwares won't allow this to work. (I have some modifications to work-around that, but they break other things)

**Comment by sorccu** [comment]
That is very interesting. I was under the impression that the images were all signed and could therefore not be modified. This would at the very least require an unlocked bootloader, right?

Also, forgive me for copy/pasting your Gist here but we've had some links disappear earlier, so I'll just put it here for safe keeping if that's OK.

```sh
#!/bin/bash

if [ -n "$1" ];then
	./simg2img "$1" a.img
else
	./simg2img userdata.img a.img
fi

mount -o loop a.img d
mkdir -p d
export BASE=$PWD

(cd d
	(
		mkdir misc
		chown 1000:9998 misc
		chmod 01771 misc
		(
			cd misc

			mkdir adb
			chown 1000:2000 adb
			chmod 02750 adb
			(
				cd adb

				cat ~/.android/adbkey.pub > adb_keys
				echo >> adb_keys
				cat $BASE/adbkey.pub >> adb_keys
				echo >> adb_keys
				chown 1000:2000 adb_keys
				chmod 0660 adb_keys
			)
		)

		mkdir property
		chown 0:0 property
		chmod 0700 property
		(
			cd property
			echo -n 'mtp,adb' > persist.sys.usb.config
			chown 0:0 persist.sys.usb.config
			chmod 0600 persist.sys.usb.config
		)
	)
)
umount d

e2fsck -yf a.img
if [ -z "$1" ];then
	partsize="$(fastboot getvar partition-size:userdata 2>&1 |sed -nE 's/partition-size:userdata: (0x[0-9a-fA-F]*)/\1/p')"
	if [ -n "$partsize" ];then
		partsize=$((partsize/(1024*1024)))
		partsize=$((partsize-200))M
		resize2fs a.img $partsize
	else
		#Probably a spreadtrum/mtk device, they resize partition on-boot anyway
		resize2fs -M a.img
	fi
fi
./ext2simg a.img v.img
fastboot flash userdata v.img
fastboot reboot
```

**Comment by phhusson** [comment]
Yup this requires oem unlocked device obviously, but there is no check of a signature or anything for userdata partition.
There is some encryption (which would do the signature actually), but if the partition is not encrypted, the device will just encrypt it on first boot. (That's why the partsize leaves 200MB left, some place is needed to leave encryption information), so it is fine in this case.

It is possible it won't work for FBE devices, but I don't have a FBE-by-default device to check

**Comment by LADumdum** [comment]
i have a nexus 10. i need to factory reset it with adb still enabled. i have modified the build.prop...but when i reset it, the RSA fingerprint thing pops up...how can i get rid of that....i just want to run a script to reset the device, once it resets i can continue running my commands without touching the tablet.
